### PR TITLE
Implement "Goto Implementation"

### LINKED ISSUE #22: Implement "Goto Implementation"
We should implement the [Goto Implementation Request](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_implementation) so that a user can navigate from a Gherkin step to the matching step definition.

This would involve (roughly):
* Refactor `ExpressionBuilder#build` to return an array of `ExpressionWithLocation`
* Return the locations of all matching expressions (`.java` or `.ts` source code), keeping in mind that it's possible for a step to match multiple expressions (ambiguous match).

### PR DESCRIPTION
**Pull request by richardruiter**
Fixes #22
WIP

### 🏷️ What kind of change is this?

- :bank: Refactoring/debt/DX (improvement to code design, tooling, documentation etc. without changing behaviour)
- :bug: Bug fix (non-breaking change which fixes a defect)
- :zap: New feature (non-breaking change which adds new behaviour)
- :boom: Breaking change (incompatible changes to the API)

### 📋 Checklist:

- [x] I agree to respect and uphold the [Cucumber Community Code of Conduct](https://cucumber.io/conduct/)
- [ ] I've changed the behaviour of the code
  - [ ] I have added/updated tests to cover my changes.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly.
- [ ] Users should know about my change
  - [ ] I have added an entry to the "Unreleased" section of the [**CHANGELOG**](../CHANGELOG.md), linking to this pull request.

----

*This text was originally generated from a [template](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/about-issue-and-pull-request-templates), then edited by hand. [You can modify the template here.](https://github.com/cucumber/.github/edit/main/.github/PULL_REQUEST_TEMPLATE.md)*

### CODE REVIEW

File: src/tree-sitter/ExpressionBuilder.ts [hunk 1]
```diff
@@ -130,12 +131,25 @@ function toString(s: string): string {
 function recordFromMatch<T extends string>(
```
Comment (aslakhellesoy) [inline src/tree-sitter/ExpressionBuilder.ts:131]: I suggest renaming the function to `glueDefinitionFromMatch`, and `RecordLink` to `GlueDefinition`

File: src/tree-sitter/ExpressionBuilder.ts [hunk 2]
```diff
@@ -75,8 +76,8 @@ export class ExpressionBuilder {
         const matches = query.matches(tree.rootNode)
         const records = matches.map((match) => recordFromMatch(match, defineParameterTypeKeys))
```
Comment (aslakhellesoy) [inline src/tree-sitter/ExpressionBuilder.ts:77]: Rename to `glueDefinitions`

File: src/tree-sitter/types.ts [hunk 3]
```diff
@@ -33,3 +34,8 @@ export interface ParserAdapter {
   setLanguage(language: LanguageName): void
   query(source: string): Parser.Query
 }
+
+export type RecordLink = {
```
Comment (aslakhellesoy) [inline src/tree-sitter/types.ts:38]: I suggest renaming this to `GlueDefinition`

Comment (aslakhellesoy) [reply src/tree-sitter/types.ts:38]: Also, make it generic for stronger type safety:

```typescript
export type GlueDefinition<T extends string> = {
   record: Record<T, string | undefined>
   link: LocationLink
}
```

**Review by aslakhellesoy — CHANGES_REQUESTED** [review]
I know this is till WIP, so I only have a little bit of feedback for now.

File: src/tree-sitter/ExpressionBuilder.ts [hunk 4]
```diff
@@ -130,12 +131,25 @@ function toString(s: string): string {
...
-  const values = keys.map((name) => match.captures.find((c) => c.name === name)?.node?.text)
-  return Object.fromEntries(keys.map((_, i) => [keys[i], values[i]])) as Record<
+): RecordLink {
+  const recordValues = keys.map((name) => match.captures.find((c) => c.name === name)?.node?.text)
+  const record = Object.fromEntries(keys.map((_, i) => [keys[i], recordValues[i]])) as Record<
     T,
     string | undefined
   >
+
+  // const linkValues = match.captures
```
Comment (richardruiter) [inline src/tree-sitter/ExpressionBuilder.ts:143]: I need some explanation how this works and how I can be sure to always match the right step definitions (also the ones with expression.

Perhaps you can right the test with proper data and I will finish the code?

Comment (aslakhellesoy) [reply src/tree-sitter/ExpressionBuilder.ts:143]: Yes, I'll have a look into this.

**Comment by aslakhellesoy** [comment]
Superseded by #48

**Pull request closed without merging**